File: android/src/main/kotlin/com/zhecent/gromore_ads_kit/utils/AdEventHelper.kt
package com.zhecent.gromore_ads_kit.utils

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import com.zhecent.gromore_ads_kit.common.AdConstants

/**
 * 广告事件处理工具类
 * 负责统一处理广告事件的发送和管理
 */
class AdEventHelper(private var eventSink: EventChannel.EventSink?) {

    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * 发送广告事件
     * @param action 事件动作
     * @param posId 广告位ID
     * @param extra 额外参数
     */
    fun sendAdEvent(action: String, posId: String, extra: Map<String, Any>? = null) {
        val event = mutableMapOf<String, Any>(
            "action" to action,
            "posId" to posId,
            "timestamp" to System.currentTimeMillis()
        )
        extra?.let {
            if (it.isNotEmpty()) {
                event["extra"] = it
            }
        }

        mainHandler.post {
            eventSink?.success(event)
        }
    }

    /**
     * 发送广告加载成功事件
     */
    fun sendLoadSuccessEvent(adType: String, posId: String, extra: Map<String, Any>? = null) {
        val action = when (adType) {
            AdConstants.AD_TYPE_SPLASH -> AdConstants.Events.SPLASH_LOADED
            AdConstants.AD_TYPE_INTERSTITIAL -> AdConstants.Events.INTERSTITIAL_LOADED
            AdConstants.AD_TYPE_REWARD_VIDEO -> AdConstants.Events.REWARD_VIDEO_LOADED
            AdConstants.AD_TYPE_FEED -> AdConstants.Events.FEED_LOADED
            AdConstants.AD_TYPE_DRAW_FEED -> AdConstants.Events.DRAW_FEED_LOADED
            AdConstants.AD_TYPE_BANNER -> AdConstants.Events.BANNER_LOADED
            else -> "${adType}_loaded"
        }
        sendAdEvent(action, posId, extra)
    }

    /**
     * 发送广告展示事件
     */
    fun sendShowEvent(adType: String, posId: String, extra: Map<String, Any>? = null) {
        val action = when (adType) {
            AdConstants.AD_TYPE_SPLASH -> AdConstants.Events.SPLASH_SHOWED
            AdConstants.AD_TYPE_INTERSTITIAL -> AdConstants.Events.INTERSTITIAL_SHOWED
            AdConstants.AD_TYPE_REWARD_VIDEO -> AdConstants.Events.REWARD_VIDEO_SHOWED
            AdConstants.AD_TYPE_FEED -> AdConstants.Events.FEED_SHOWED
            AdConstants.AD_TYPE_DRAW_FEED -> AdConstants.Events.DRAW_FEED_SHOWED
            AdConstants.AD_TYPE_BANNER -> AdConstants.Events.BANNER_SHOWED
            else -> "${adType}_showed"
        }
        sendAdEvent(action, posId, extra)
    }

    /**
     * 发送广告点击事件
     */
    fun sendClickEvent(adType: String, posId: String, extra: Map<String, Any>? = null) {
        val action = when (adType) {
            AdConstants.AD_TYPE_SPLASH -> AdConstants.Events.SPLASH_CLICKED
            AdConstants.AD_TYPE_INTERSTITIAL -> AdConstants.Events.INTERSTITIAL_CLICKED
            AdConstants.AD_TYPE_REWARD_VIDEO -> AdConstants.Events.REWARD_VIDEO_CLICKED
            AdConstants.AD_TYPE_FEED -> AdConstants.Events.FEED_CLICKED
            AdConstants.AD_TYPE_DRAW_FEED -> AdConstants.Events.DRAW_FEED_CLICKED
            AdConstants.AD_TYPE_BANNER -> AdConstants.Events.BANNER_CLICKED
            else -> "${adType}_clicked"
        }
        sendAdEvent(action, posId, extra)
    }

    /**
     * 发送广告关闭事件
     */
    fun sendCloseEvent(adType: String, posId: String, extra: Map<String, Any>? = null) {
        val action = when (adType) {
            AdConstants.AD_TYPE_SPLASH -> AdConstants.Events.SPLASH_CLOSED
            AdConstants.AD_TYPE_INTERSTITIAL -> AdConstants.Events.INTERSTITIAL_CLOSED
            AdConstants.AD_TYPE_REWARD_VIDEO -> AdConstants.Events.REWARD_VIDEO_CLOSED
            AdConstants.AD_TYPE_FEED -> AdConstants.Events.FEED_CLOSED
            AdConstants.AD_TYPE_DRAW_FEED -> AdConstants.Events.DRAW_FEED_CLOSED
            AdConstants.AD_TYPE_BANNER -> AdConstants.Events.BANNER_CLOSED
            else -> "${adType}_closed"
        }
        sendAdEvent(action, posId, extra)
    }

    /**
     * 发送广告加载失败事件
     */
    fun sendLoadFailEvent(adType: String, posId: String, errorCode: Int, errorMsg: String) {
        val action = when (adType) {
            AdConstants.AD_TYPE_SPLASH -> AdConstants.Events.SPLASH_LOAD_FAIL
            AdConstants.AD_TYPE_INTERSTITIAL -> AdConstants.Events.INTERSTITIAL_LOAD_FAIL
            AdConstants.AD_TYPE_REWARD_VIDEO -> AdConstants.Events.REWARD_VIDEO_LOAD_FAIL
            AdConstants.AD_TYPE_FEED -> AdConstants.Events.FEED_LOAD_FAIL
            AdConstants.AD_TYPE_DRAW_FEED -> AdConstants.Events.DRAW_FEED_LOAD_FAIL
            AdConstants.AD_TYPE_BANNER -> AdConstants.Events.BANNER_LOAD_FAIL
            else -> "${adType}_load_fail"
        }
        sendAdEvent(action, posId, mapOf(
            "message" to errorMsg,
            "code" to errorCode
        ))
    }

    /**
     * 发送Banner专用事件
     */
    fun sendBannerEvent(action: String, posId: String, extra: Map<String, Any>? = null) {
        sendAdEvent(action, posId, extra)
    }

    /**
     * 发送Banner错误事件
     */
    fun sendBannerErrorEvent(action: String, posId: String, errorCode: Int, errorMsg: String) {
        sendAdEvent(action, posId, mapOf(
            "message" to errorMsg,
            "code" to errorCode,
            "adType" to AdConstants.AD_TYPE_BANNER
        ))
    }

    /**
     * 发送Banner ECPM事件
     */
    fun sendBannerEcpmEvent(posId: String, ecpmData: Map<String, Any>) {
        sendAdEvent(AdConstants.Events.BANNER_ECPM, posId, ecpmData)
    }

    /**
     * 更新事件通道
     */
    fun updateEventSink(eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink
    }

    companion object {
        @Volatile
        private var INSTANCE: AdEventHelper? = null

        /**
         * 获取单例实例
         */
        fun getInstance(eventSink: EventChannel.EventSink? = null): AdEventHelper {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: AdEventHelper(eventSink).also { INSTANCE = it }
            }
        }
    }
}

File: ios/Classes/Utils/AdEventHelper.swift
import Foundation
import Flutter

/**
 * 广告事件助手类
 * 负责事件的统一发送和管理
 */
class AdEventHelper {

    // 单例实例
    static let shared = AdEventHelper()

    // 事件通道回调
    private var eventSink: FlutterEventSink?

    private init() {}

    /**
     * 更新事件回调
     */
    func updateEventSink(_ sink: FlutterEventSink?) {
        eventSink = sink
        NSLog("\(AdConstants.TAG) 事件通道已更新: \(sink != nil ? "已连接" : "已断开")")
    }

    /**
     * 发送广告事件
     */
    func sendAdEvent(_ eventType: String, posId: String, extra: [String: Any]? = nil) {
        guard let eventSink = eventSink else {
            NSLog("\(AdConstants.TAG) 事件通道未连接，无法发送事件: \(eventType)")
            return
        }

        var eventData: [String: Any] = [
            "action": eventType,
            "posId": posId,
            "timestamp": Date().timeIntervalSince1970 * 1000
        ]

        // 合并额外数据
        if let extra = extra {
            if !extra.isEmpty {
                eventData["extra"] = extra
            }
        }

        // 在主线程发送事件
        DispatchQueue.main.async {
            eventSink(eventData)
            NSLog("\(AdConstants.TAG) 发送事件: \(eventType), posId: \(posId)")
        }
    }

    /**
     * 发送开屏广告事件
     */
    func sendSplashEvent(_ eventType: String, posId: String, extra: [String: Any]? = nil) {
        sendAdEvent(eventType, posId: posId, extra: extra)
    }

    /**
     * 发送插屏广告事件
     */
    func sendInterstitialEvent(_ eventType: String, posId: String, extra: [String: Any]? = nil) {
        sendAdEvent(eventType, posId: posId, extra: extra)
    }

    /**
     * 发送激励视频事件
     */
    func sendRewardVideoEvent(_ eventType: String, posId: String, extra: [String: Any]? = nil) {
        sendAdEvent(eventType, posId: posId, extra: extra)
    }

    /**
     * 发送信息流广告事件
     */
    func sendFeedEvent(_ eventType: String, posId: String, extra: [String: Any]? = nil) {
        sendAdEvent(eventType, posId: posId, extra: extra)
    }

    /**
     * 发送展示事件
     */
    func sendShowEvent(_ adType: String, posId: String, extra: [String: Any]? = nil) {
        let action = "\(adType)_showed"
        sendAdEvent(action, posId: posId, extra: extra)
    }

    /**
     * 发送点击事件
     */
    func sendClickEvent(_ adType: String, posId: String, extra: [String: Any]? = nil) {
        let action = "\(adType)_clicked"
        sendAdEvent(action, posId: posId, extra: extra)
    }

    /**
     * 发送关闭事件
     */
    func sendCloseEvent(_ adType: String, posId: String, extra: [String: Any]? = nil) {
        let action = "\(adType)_closed"
        sendAdEvent(action, posId: posId, extra: extra)
    }

    /**
     * 发送Banner广告事件
     */
    func sendBannerEvent(_ eventType: String, posId: String, extra: [String: Any]? = nil) {
        sendAdEvent(eventType, posId: posId, extra: extra)
    }

    /**
     * 发送错误事件
     */
    func sendErrorEvent(adType: String, posId: String, errorCode: Int, errorMessage: String) {
        let eventType = "\(adType)_load_fail"
        let extra: [String: Any] = [
            "code": errorCode,
            "message": errorMessage,
            "adType": adType
        ]
        sendAdEvent(eventType, posId: posId, extra: extra)
    }

    /**
     * 发送Banner错误事件
     */
    func sendBannerErrorEvent(_ eventType: String, message: String, posId: String, code: Int? = nil) {
        var extra: [String: Any] = [
            "message": message,
            "adType": "banner"
        ]

        if let errorCode = code {
            extra["code"] = errorCode
        }

        sendAdEvent(eventType, posId: posId, extra: extra)
    }
}
